stl/inc/memory

```diff
@@ -1522,6 +1562,280 @@ private:
+class _Ref_count_unbounded_array : public _Ref_count_base {
+    // handle reference counting for unbounded array with trivial destruction in control block, no allocator
+public:
+    using _Element_type = remove_extent_t<_Ty>;
+
+    explicit _Ref_count_unbounded_array(const size_t _Count) : _Ref_count_base() {
+        _Uninitialized_value_construct_multidimensional_n(_Get_ptr(), _Count);
+    }
+
+    explicit _Ref_count_unbounded_array(const size_t _Count, const _Element_type& _Val) : _Ref_count_base() {
+        _Uninitialized_fill_multidimensional_n(_Get_ptr(), _Count, _Val);
+    }
+
+    _NODISCARD auto _Get_ptr() noexcept {
+        return _STD addressof(_Storage._Value);
+    }
+
+private:
+    union {
+        _Wrap<_Element_type> _Storage; // flexible array must be last member
+    };
+
+    ~_Ref_count_unbounded_array() {
+        // nothing to do, _Ty is trivially destructible
+
+        // See N4849 [class.dtor]/7.
+    }
+
+    virtual void _Destroy() noexcept override { // destroy managed resource
+        // nothing to do, _Ty is trivially destructible
+    }
+
+    virtual void _Delete_this() noexcept override { // destroy self
+        this->~_Ref_count_unbounded_array();
+        ::operator delete(static_cast<void*>(this));
+    }
+};
+
+template <class _Ty>
+class _Ref_count_unbounded_array<_Ty, false> : public _Ref_count_base {
+    // handle reference counting for unbounded array with non-trivial destruction in control block, no allocator
+public:
+    using _Element_type = remove_extent_t<_Ty>;
+
+    explicit _Ref_count_unbounded_array(const size_t _Count) : _Ref_count_base(), _Size(_Count) {
+        _Uninitialized_value_construct_multidimensional_n(_Get_ptr(), _Size);
+    }
+
+    explicit _Ref_count_unbounded_array(const size_t _Count, const _Element_type& _Val)
+        : _Ref_count_base(), _Size(_Count) {
+        _Uninitialized_fill_multidimensional_n(_Get_ptr(), _Size, _Val);
+    }
+
+    _NODISCARD auto _Get_ptr() noexcept {
+        return _STD addressof(_Storage._Value);
+    }
+
+private:
+    size_t _Size;
+
+    union {
+        _Wrap<_Element_type> _Storage; // flexible array must be last member
+    };
+
+    ~_Ref_count_unbounded_array() {
+        // nothing to do, _Storage was already destroyed in _Destroy
+
+        // See N4849 [class.dtor]/7.
+    }
+
+    virtual void _Destroy() noexcept override { // destroy managed resource
+        _Reverse_destroy_multidimensional_n(_Get_ptr(), _Size);
+    }
+
+    virtual void _Delete_this() noexcept override { // destroy self
+        this->~_Ref_count_unbounded_array();
+        ::operator delete(static_cast<void*>(this));
+    }
+};
+
+// CLASS TEMPLATE _Ref_count_bounded_array
+template <class _Ty>
+class _Ref_count_bounded_array : public _Ref_count_base {
+    // handle reference counting for bounded array in control block, no allocator
+public:
+    _Ref_count_bounded_array() : _Ref_count_base(), _Storage() {}

```

Comment:
Doesn't seem like we need to value-initialize `_Storage` here.

Reply:
This is actually necessary. Without it, `make_shared<string[5]>()` would contain garbage. (Alternatively, we could avoid value-initialization and immediately call `_Uninitialized_value_construct_multidimensional_n`, but that's probably going to result in worse codegen.)

I've added a comment drawing attention to the necessary value-initialization here. I've also added comments to the other constructors where we skip value-init, because adding an apparently-innocuous `_Storage()` to them would be bad.